File: src/typegraph/rust/typegraph/src/runtimes/kv.rs
// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

use crate::{
    t::{self, TypeDef},
    wasm::{
        self,
        core::{MaterializerId, RuntimeId},
        runtimes::{BaseMaterializer, Effect, KvMaterializer, KvRuntimeData},
    },
    Result,
};

#[derive(Debug)]
pub struct KvRuntime {
    id: RuntimeId,
}

impl KvRuntime {
    pub fn new(url: &str) -> Result<Self> {
        let data = KvRuntimeData {
            url: url.to_string(),
        };

        let id = wasm::with_runtimes(|r, s| r.call_register_kv_runtime(s, &data))?;

        Ok(Self { id })
    }

    pub fn set(&self) -> Result<TypeDef> {
        let inp = t::structb! {
            "key" => t::string(),
            "value" => t::string(),
        };
        let out = t::string();
        let mat = self.operation(KvMaterializer::Set, Effect::Update(false))?;

        t::funcb(inp, out, mat)
    }

    pub fn get(&self) -> Result<TypeDef> {
        let inp = t::structb! { "key" => t::string() };
        let out = t::string();
        let mat = self.operation(KvMaterializer::Get, Effect::Read)?;

        t::funcb(inp, out, mat)
    }

    pub fn delete(&self) -> Result<TypeDef> {
        let inp = t::structb! { "key" => t::string() };
        let out = t::integer();
        let mat = self.operation(KvMaterializer::Delete, Effect::Delete(false))?;

        t::funcb(inp, out, mat)
    }

    pub fn keys(&self) -> Result<TypeDef> {
        let inp = t::structb! { "filter" => t::string().optional() };
        let out = t::list(t::string());
        let mat = self.operation(KvMaterializer::Keys, Effect::Read)?;

        t::funcb(inp, out, mat)
    }

    pub fn values(&self) -> Result<TypeDef> {
        let inp = t::structb! { "filter" => t::string().optional() };
        let out = t::list(t::string());
        let mat = self.operation(KvMaterializer::Values, Effect::Read)?;

        t::funcb(inp, out, mat)
    }

    fn operation(&self, kind: KvMaterializer, effect: Effect) -> Result<MaterializerId> {
        let base = BaseMaterializer {
            runtime: self.id,
            effect,
        };

        wasm::with_runtimes(|r, s| r.call_kv_operation(s, base, kind))
    }
}

File: src/typegraph/python/typegraph/runtimes/kv.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from dataclasses import dataclass

from typegraph import fx, t
from typegraph.gen.runtimes import (
    BaseMaterializer,
    Effect,
    KvMaterializer,
    KvRuntimeData,
)
from typegraph.runtimes.base import Materializer, Runtime
from typegraph.sdk import runtimes


class KvRuntime(Runtime):
    url: str

    def __init__(self, url: str):
        data = KvRuntimeData(url)
        runtime_id = runtimes.register_kv_runtime(data)
        super().__init__(runtime_id)
        self.url = url

    def get(self):
        mat = self.__operation("get", fx.read())
        return t.func(t.struct({"key": t.string()}), t.string().optional(), mat)

    def set(self):
        mat = self.__operation("set", fx.update())
        return t.func(
            t.struct({"key": t.string(), "value": t.string()}), t.string(), mat
        )

    def delete(self):
        mat = self.__operation("delete", fx.delete())
        return t.func(t.struct({"key": t.string()}), t.integer(), mat)

    def keys(self):
        mat = self.__operation("keys", fx.read())
        return t.func(
            t.struct({"filter": t.optional(t.string())}), t.list(t.string()), mat
        )

    def values(self):
        mat = self.__operation("values", fx.read())
        return t.func(
            t.struct({"filter": t.optional(t.string())}),
            t.list(t.string()),
            mat,
        )

    def lpush(self):
        mat = self.__operation("lpush", fx.update())
        return t.func(
            t.struct({"key": t.string(), "value": t.string()}), t.integer(), mat
        )

    def rpush(self):
        mat = self.__operation("rpush", fx.update())
        return t.func(
            t.struct({"key": t.string(), "value": t.string()}), t.integer(), mat
        )

    def lpop(self):
        mat = self.__operation("lpop", fx.update())
        return t.func(
            t.struct({"key": t.string()}),
            t.optional(t.string()),
            mat,
        )

    def rpop(self):
        mat = self.__operation("rpop", fx.update())
        return t.func(
            t.struct({"key": t.string()}),
            t.optional(t.string()),
            mat,
        )

    def __operation(self, operation: KvMaterializer, effect: Effect):
        mat_id = runtimes.kv_operation(BaseMaterializer(self.id, effect), operation)

        return KvOperationMat(mat_id, effect=effect, operation=operation)


@dataclass
class KvOperationMat(Materializer):
    operation: KvMaterializer
